Extract consensus logic into separate target

makadaw opened the pull request:
Make consensus logic independent from threads and RPC. `Consensus` type
describe the Raft protocol logic and do not control time or
communication.
Such split make tests of the protocol trivial and ability to implement
communication layer separately.

makadaw committed:
Extract consensus logic into separate target

Make consensus logic independent from threads and RPC. `Consensus` type
describe the Raft protocol logic and do not control time or
communication.
Such split make tests of the protocol trivial and ability to implement
communication layer separately.

ktoso commented on Sources/Raft/Consensus/Consensus.swift:
```diff
@@ -0,0 +1,227 @@
+import enum Dispatch.DispatchTimeInterval
+import Logging
+
+actor public class Consensus {
```
probably 
```suggestion
public actor class Consensus {
```

as a heads up, this soon will be just `public actor`

makadaw replied:
Yes, with last snapshot (16-02) just `public actor` works

ktoso commented on Sources/Raft/Consensus/Consensus.swift:
```diff
@@ -0,0 +1,227 @@
+    func electionTimeoutInterval() -> DispatchTimeInterval {
+        .nanoseconds(Int(config.protocol.electionTimeout.nanoseconds
+                            + Int64.random(in: 1000...config.protocol.electionTimeout.nanoseconds)))
+    }
```
Good, using `DispatchTimeInterval` is also a good choice. 

Sadly it's the best we can do for the time being, until swift tackles adding some of it's own time types

ktoso commented on Sources/Raft/Consensus/Consensus.swift:
```diff
@@ -0,0 +1,227 @@
+
+    func electionTimeoutInterval() -> DispatchTimeInterval {
+        .nanoseconds(Int(config.protocol.electionTimeout.nanoseconds
+                            + Int64.random(in: 1000...config.protocol.electionTimeout.nanoseconds)))
```
might want to stuff such calculation into the config/settings object (the entire func)

ktoso commented on Sources/Raft/Consensus/Consensus.swift:
```diff
@@ -0,0 +1,227 @@
+    }
+
+    /// Current node id
+    var myself: NodeId {
```
very minor nitpick, but generally we spell such as `...ID`, just in case you want to be consistent with other libs

makadaw replied:
Don't have any preferences here, but better to be consistent. Will update

ktoso commented on Sources/Raft/Consensus/Consensus.swift:
```diff
@@ -0,0 +1,227 @@
+        }
+        return false
+    }
+}
```
👍

ktoso commented on Sources/Raft/Consensus/Consensus.swift:
```diff
@@ -0,0 +1,227 @@
+        self.config = config
+        self.peers = peers
+
+        self.term = Term(myself: config.myself.id)
```
👍

ktoso commented on Sources/Raft/Consensus/Consensus.swift:
```diff
@@ -0,0 +1,227 @@
+                return false
+            } catch {
+                // If got any error we lost an election
+                return false
```
minor thing but you could `group.cancelAll()` before returning here... It won't really be able to do much if the requests are already shot off, but if they're not the tasks are cancelled and we can avoid unnecessary work

makadaw replied:
Will also add cancellation to the case when we already got enough voces

ktoso commented on Sources/Raft/Consensus/Consensus.swift:
```diff
@@ -0,0 +1,227 @@
+            let voteGranted: Bool
+        }
+    }
+
```
Looks like a good start 👍

ktoso commented on Sources/Raft/Consensus/Consensus.swift:
```diff
@@ -0,0 +1,227 @@
+            "vote/self": "\(myself)",
+            "vote/candidate": "\(request.candidate)",
+            "vote/granted": "\(granted)"
+        ])
```
good logging 👍 +1 for using metadata like this, it's a great pattern to follow

ktoso commented on Sources/Raft/Consensus/Consensus.swift:
```diff
@@ -0,0 +1,227 @@
+}
+
+// MARK: Entries
+extension Consensus {
```
I guess I would call this `Raft` or something else perhaps? Consensus is a bit weird name for it hm

ktoso replied:
(it'd conflict with the module name, but generally we try to prefer nice type names, and less nice module names; so `SwiftRaft` could be a nice module name etc)

makadaw replied:
Before I have used `ConsensusService`, but it also looks wired and naming is hard :) Will update module name to `SwiftRaft` and actor to `Raft`

ktoso commented on Sources/Raft/Consensus/State.swift:
```diff
@@ -0,0 +1,32 @@
+        default:
+            return false
+        }
+    }
```
👍

ktoso commented on Sources/Raft/Peer.swift:
```diff
@@ -0,0 +1,15 @@
+    var quorumSize: UInt {
+        UInt(count/2 + 1)
+    }
+}
```
Oh the number of times I did that exact same extension on collections... 😆

ktoso reviewed (approved):
Looks like a great start 👍 

Great to extract the core of it to a separate module.

makadaw committed:
Rename Id in types to ID

makadaw committed:
Move election timeout calculation into configuration

makadaw committed:
Migrate to the last Swift toolchain

Replace `actor class` to just `actor`.
Make a replacement for deprecated run `runAsyncAndBlock` in unit tests.

makadaw committed:
Rename module to SwiftRaft

makadaw reviewed (commented):
@ktoso Thanks for the review!

makadaw committed:
Delete unused file
